src/security/content-security-policy.conf

```diff
@@ -8,18 +8,66 @@
+# The example policy below aims to:
 #
+#  (1) Restrict all fetches by default to the origin of the current website
+#      by setting the `default-src` directive to `'self'` - which acts as a
+#      fallback to all "Fetch directives" (https://developer.mozilla.org/en-US/docs/Glossary/Fetch_directive).
+#
+#      This is conveniant as you do not have to specify all Fetch directives
+#      that apply to your site, for example:
+#      `connect-src 'self'; font-src 'self'; script-src 'self'; style-src 'self'`, etc.
+#
+#      This restriction also means that you must explicitly define from
+#      which site(s) your website is allowed to load resources from.
+#
+#  (2) The `<base>` element is not allowed on the website. This is to
+#      prevent attackers from changing the locations of resources loaded
+#      from relative URLs.
+#
+#      If you want to use the `<base>` element, then `base-uri 'self'`
+#      can be used instead.
+#
+#  (3) Form submissions are only allowed from the current website by
+#      setting: `form-action 'self'`.
+#
+#  (4) Prevents all websites (including your own) from embedding your
+#      webpages within e.g. the `<iframe>` or `<object>` element by
+#      setting `frame-ancestors 'none'`.
+#
+#	   The `frame-ancestors` directive helps avoid "Clickjacking" attacks
+#      and is similar to the `X-Frame-Options` header.
+#
+#      Browsers that support the CSP header will ignore `X-Frame-Options`
+#      if `frame-ancestors` is also specified.
+#
+#  (5) Forces the browser to treat all the resources that are served over
+#      HTTP as if they were loaded securely over HTTPS by setting the
+#      `upgrade-insecure-requests` directive.
+#
+#      Please note that `upgrade-insecure-requests` does not ensure
+#      HTTPS for the top-level navigation. If you want to force the
+#      website itself to be loaded over HTTPS you must include the
+#      `Strict-Transport-Security` header.
+#
+# To make your CSP implementation easier, you can use an online CSP header
+# generator such as:
+# https://report-uri.com/home/generate/
+#
+# It is encouraged that you validate your CSP header using a CSP validator
+# such as:
+# https://csp-evaluator.withgoogle.com
+#
+# https://csp.withgoogle.com/docs/
 # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
-# https://www.w3.org/TR/CSP3/
-# https://content-security-policy.com/
 # https://www.html5rocks.com/en/tutorials/security/content-security-policy/
+# https://www.w3.org/TR/CSP/
 
 <IfModule mod_headers.c>
-    Header set Content-Security-Policy "script-src 'self'; object-src 'self'" "expr=%{CONTENT_TYPE} =~ m#text/html#i"
+    #                                   (1)                 (2)              (3)                 (4)                     (5)
+    Header set Content-Security-Policy "default-src 'self'; base-uri 'none'; form-action 'self'; frame-ancestors 'none'; upgrade-insecure-requests" "expr=%{CONTENT_TYPE} =~ m#text/html#i"
```

Comment:
Suggestion from https://github.com/h5bp/server-configs-nginx/pull/224#issuecomment-476101646:

```suggestion
    Header set Content-Security-Policy "default-src 'none'; frame-ancestors 'none'; form-action 'none'; upgrade-insecure-requests 1; block-all-mixed-content; disown-opener; sandbox; reflected-xss block; base-uri 'none'; referrer no-referrer-when-downgrade" "expr=%{CONTENT_TYPE} =~ m#text/html#i"
```

Basically:
* Shouldn't we use `none` for `default-src`? This might be to restrictive for scripts and styles, but more secure for all others.
* `block-all-mixed-content`, `disown-opener` and `sandbox` seem safe enough but are they useful?
* `reflected-xss block` modern way to [`X-XSS-Protection`](https://github.com/h5bp/server-configs-apache/blob/e12ae05c4ae690e36386de0efc2cfc0186f2f4f2/src/security/x-xss-protection.conf).
* `referrer no-referrer-when-downgrade` alternative to [`Referrer-Policy`](https://github.com/h5bp/server-configs-apache/blob/e12ae05c4ae690e36386de0efc2cfc0186f2f4f2/src/security/referrer-policy.conf).

Thoughts?

Reply:
I think we should keep this simple, at least for starters.

So I wouldn't duplicate the directives that are covered by others like Referrer-Policy etc.

About default-src, while I agree, it's the issue we discussed. Many other CSP directives will inherit this. Perhaps `'self'` is a more sane default.

About block-all-mixed-content, disown-opener and sandbox, not sure I'd add them right now. But this should be revisited later for sure.

Reply:
I agree with @XhmikosR keeping `default-src` to `'self'` mainly because it is an easier implementation model than specifying each and every fetch directive used by the application.

`sandbox` is very restrictive and developers must be aware of which Web APIs are being used in their application then allow them accordingly. Many of these [values](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/sandbox#Syntax) would have to be enabled again anyways to preserve normal web behaviors, such as `allow-forms`, `allow-scripts`, `allow-top-navigation`. So just having other CSP directives allowing _who_ can use things instead of `sandbox` dictating _what_ seems like the best default. Developers who want advanced and super strict policies certainly will look closer at directives such as `sandbox`.

`disown-opener` seems to be deprecated: https://github.com/w3c/webappsec-csp/issues/194. I think a new HTTP header [`Cross-Origin-Opener-Policy`](https://github.com/whatwg/html/issues/3740) is to replace it.

[According to MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/referrer), `referrer` is obsolete and we should stick to the `Referrer-Policy` header instead.

`block-all-mixed-content` is similar to `upgrade-insecure-requests` and if the latter is specified then the `block-all-mixed-content` directive is ignored. Although `block-all-mixed-content` has a tiny bit wider support I think we should default to `upgrade-insecure-requests` which [Google describes](https://developers.google.com/web/fundamentals/security/prevent-mixed-content/fixing-mixed-content#upgrading_insecure_requests) as "the newest and best tools to automatically fix mixed content" - the directive simply upgrades HTTP to HTTPS while `block-all-mixed-content` (seems to me) to straight up block HTTP requests and Google describe that directive with: "The downside of using `block-all-mixed-content` is, perhaps obviously, that all content is blocked."

> `reflected-xss block` modern way to [`X-XSS-Protection`](https://github.com/h5bp/server-configs-apache/blob/e12ae05c4ae690e36386de0efc2cfc0186f2f4f2/src/security/x-xss-protection.conf).

I don't know what this is, I've not been able to find such a CSP directive defined anywhere.

Reply:
Thanks a lot for feedbacks!
And well done, @Malvoz, very clear 👍 

> > `reflected-xss block` modern way to [`X-XSS-Protection`](https://github.com/h5bp/server-configs-apache/blob/e12ae05c4ae690e36386de0efc2cfc0186f2f4f2/src/security/x-xss-protection.conf).
> 
> I don't know what this is

Seems deprecated too: https://www.chromestatus.com/feature/5769374145183744

All good then!